File: behavioral/chain-of-responsibility/python/chain_of_responsibility.py
class Handler:
    def __init__(self):
        self._next_handler = None

    def set_next(self, handler):
        self._next_handler = handler

    def handle_request(self, request):
        raise NotImplementedError("This method should be overridden by subclasses")


class ConcreteHandler1(Handler):
    def handle_request(self, request):
        if request == "Request1":
            print("ConcreteHandler1 handled the request.")
        elif self._next_handler is not None:
            self._next_handler.handle_request(request)


class ConcreteHandler2(Handler):
    def handle_request(self, request):
        if request == "Request2":
            print("ConcreteHandler2 handled the request.")
        elif self._next_handler is not None:
            self._next_handler.handle_request(request)


if __name__ == "__main__":
    handler1 = ConcreteHandler1()
    handler2 = ConcreteHandler2()

    handler1.set_next(handler2)

    handler1.handle_request("Request1")
    handler1.handle_request("Request2")
    handler1.handle_request("UnknownRequest")

File: src/behavioral/chain-of-responsibility/ChainOfResponsibility.cs
namespace DesignPatterns.Behavioral.ChainOfResponsibility;

public abstract class Handler
{
    protected Handler nextHandler;

    public void SetNext(Handler nextHandler)
    {
        this.nextHandler = nextHandler;
    }

    public abstract void HandleRequest(string request);
}

public class ConcreteHandler1 : Handler
{
    public override void HandleRequest(string request)
    {
        if (request == "Request1")
        {
            Console.WriteLine("ConcreteHandler1 handled the request.");
        }
        else if (nextHandler != null)
        {
            nextHandler.HandleRequest(request);
        }
    }
}

public class ConcreteHandler2 : Handler
{
    public override void HandleRequest(string request)
    {
        if (request == "Request2")
        {
            Console.WriteLine("ConcreteHandler2 handled the request.");
        }
        else if (nextHandler != null)
        {
            nextHandler.HandleRequest(request);
        }
    }
}

public class Program
{
    public static void Main()
    {
        Handler handler1 = new ConcreteHandler1();
        Handler handler2 = new ConcreteHandler2();

        handler1.SetNext(handler2);

        handler1.HandleRequest("Request1");
        handler1.HandleRequest("Request2");
        handler1.HandleRequest("UnknownRequest");
    }
}
